src/sinks/util/statistic.rs

```diff
@@ -21,60 +22,86 @@ pub struct DistributionStatistic {
 
 impl DistributionStatistic {
     pub fn from_samples(source: &[Sample], quantiles: &[f64]) -> Option<Self> {
-        let mut samples = Vec::new();
-        for sample in source {
-            for _ in 0..sample.rate {
-                samples.push(sample.value);
-            }
-        }
+        let mut bins = source
+            .iter()
+            .filter(|sample| sample.rate > 0)
+            .cloned()
+            .collect::<Vec<_>>();
 
-        if samples.is_empty() {
+        if bins.is_empty() {
             return None;
         }
 
-        if samples.len() == 1 {
-            let val = samples[0];
+        if bins.len() == 1 {
+            let val = bins[0].value;
+            let count = bins[0].rate;
             return Some(Self {
                 min: val,
                 max: val,
                 median: val,
                 avg: val,
-                sum: val,
-                count: 1,
+                sum: val * count as f64,
+                count: count as u64,
                 quantiles: quantiles.iter().map(|&p| (p, val)).collect(),
             });
         }
 
-        samples.sort_by(|a, b| a.partial_cmp(b).unwrap_or(Ordering::Equal));
+        bins.sort_unstable_by(|a, b| a.value.partial_cmp(&b.value).unwrap_or(Ordering::Equal));
+
+        let min = bins.first().unwrap().value;
+        let max = bins.last().unwrap().value;
+        let sum = bins
+            .iter()
+            .map(|sample| sample.value * sample.rate as f64)
+            .sum::<f64>();
+        let count = bins.iter().map(|sample| sample.rate).sum::<u32>();
+        let avg = sum / count as f64;
 
-        let length = samples.len() as f64;
-        let min = *samples.first().unwrap();
-        let max = *samples.last().unwrap();
+        for i in 1..bins.len() {
+            bins[i].rate += bins[i - 1].rate;
+        }
 
-        let median = samples[(0.50 * length - 1.0).round() as usize];
+        let median = find_quantile(&bins, 0.5);
         let quantiles = quantiles
             .iter()
-            .map(|&p| {
-                let sample = samples[(p * length - 1.0).round() as usize];
-                (p, sample)
-            })
+            .map(|&p| (p, find_quantile(&bins, p)))
```

Comment:
Has any testing been done, even micro-benchmarks, to demonstrate this is at least not worse than the previous approach of duplicating each sample N times and then extracting the quantiles?

Reply:
Not yet. It defends on the input, the new code is definitely better when counts are large. When counts are mostly 1 - 2, it will be slower due to binary search for each quantile. I could try some benchmark to see how much it is slower, and when the new code starts performing better.

If we can enforce `quantiles` to be sorted, it is possible to run one loop to calculate all quantiles, I believe this strategy is better on all cases.

Reply:
It would be nice to add benches to our benchmarking suite to answer these questions. Ideally, we'd get a sense of real-world performance changes with these benchmarks. @juchiast is it worth adding them here?

Reply:
> the new code is definitely better when counts are large

This is what I would expect as well, but it is not possible to say it is definitely better without measuring at what number of samples it becomes better, and some estimate of what kinds of distributions we handle.

> If we can enforce `quantiles` to be sorted, it is possible to run one loop to calculate all quantiles.

Do you mean something more than the `bins.sort_unstable_by` you are already doing here?

Reply:
We could sort `quantiles` here too, but it is more ideal to sort it at sink building time, and add an ordering check to `validate_quantiles`.

Reply:
Just pushed some benchmarks, samples count from 5 to 1000. Contrary to my prediction, the new code performed better on all cases.

This branch:
```
statistic/samples-5     time:   [171.17 ns 171.56 ns 171.93 ns]
Found 4 outliers among 100 measurements (4.00%)
  2 (2.00%) low severe
  1 (1.00%) low mild
  1 (1.00%) high mild
statistic/samples-10    time:   [320.93 ns 323.30 ns 325.47 ns]
statistic/samples-50    time:   [811.67 ns 814.12 ns 816.65 ns]
Found 11 outliers among 100 measurements (11.00%)
  3 (3.00%) low severe
  1 (1.00%) low mild
  1 (1.00%) high mild
  6 (6.00%) high severe
statistic/samples-100   time:   [1.5106 us 1.5142 us 1.5175 us]
Found 6 outliers among 100 measurements (6.00%)
  4 (4.00%) high mild
  2 (2.00%) high severe
statistic/samples-200   time:   [2.6074 us 2.6116 us 2.6164 us]
Found 4 outliers among 100 measurements (4.00%)
  1 (1.00%) low mild
  1 (1.00%) high mild
  2 (2.00%) high severe
statistic/samples-500   time:   [6.4158 us 6.4236 us 6.4325 us]
Found 6 outliers among 100 measurements (6.00%)
  2 (2.00%) low severe
  1 (1.00%) low mild
  2 (2.00%) high mild
  1 (1.00%) high severe
statistic/samples-1000  time:   [12.666 us 12.705 us 12.752 us]
Found 5 outliers among 100 measurements (5.00%)
  1 (1.00%) low mild
  1 (1.00%) high mild
  3 (3.00%) high severe
```

Master:
```
statistic/samples-5     time:   [182.07 ns 182.22 ns 182.44 ns]
                        change: [+6.2620% +6.9247% +7.5444%] (p = 0.00 < 0.05)
                        Performance has regressed.
Found 8 outliers among 100 measurements (8.00%)
  5 (5.00%) low severe
  2 (2.00%) low mild
  1 (1.00%) high mild
statistic/samples-10    time:   [342.15 ns 342.74 ns 343.70 ns]
                        change: [+5.5688% +6.3627% +7.1266%] (p = 0.00 < 0.05)
                        Performance has regressed.
Found 12 outliers among 100 measurements (12.00%)
  8 (8.00%) low severe
  1 (1.00%) low mild
  2 (2.00%) high mild
  1 (1.00%) high severe
statistic/samples-50    time:   [1.2330 us 1.2351 us 1.2371 us]
                        change: [+45.439% +48.280% +50.801%] (p = 0.00 < 0.05)
                        Performance has regressed.
Found 10 outliers among 100 measurements (10.00%)
  2 (2.00%) low severe
  7 (7.00%) low mild
  1 (1.00%) high severe
statistic/samples-100   time:   [2.2284 us 2.2310 us 2.2337 us]
                        change: [+46.939% +48.231% +49.264%] (p = 0.00 < 0.05)
                        Performance has regressed.
Found 9 outliers among 100 measurements (9.00%)
  1 (1.00%) low mild
  6 (6.00%) high mild
  2 (2.00%) high severe
statistic/samples-200   time:   [4.5200 us 4.5276 us 4.5343 us]
                        change: [+71.754% +72.318% +72.921%] (p = 0.00 < 0.05)
                        Performance has regressed.
Found 3 outliers among 100 measurements (3.00%)
  2 (2.00%) high mild
  1 (1.00%) high severe
statistic/samples-500   time:   [14.960 us 14.972 us 14.985 us]
                        change: [+133.56% +134.18% +134.82%] (p = 0.00 < 0.05)
                        Performance has regressed.
Found 6 outliers among 100 measurements (6.00%)
  6 (6.00%) high mild
statistic/samples-1000  time:   [36.663 us 36.687 us 36.712 us]
                        change: [+190.53% +191.98% +193.31%] (p = 0.00 < 0.05)
                        Performance has regressed.
Found 11 outliers among 100 measurements (11.00%)
  4 (4.00%) high mild
  7 (7.00%) high severe
```

Reply:
Samples are generated randomly with count of each bin sampled from [1, 2, 3]. I'll be testing more to confirm it.

Reply:
Note that some of the performance gain comes from `sort_unstable_by`.

Reply:
Ok so here is the final benchmark result:
```
distribution_statistic/small-bin-5
                        time:   [186.80 ns 188.08 ns 189.10 ns]
                        change: [-0.3713% +0.6682% +1.7541%] (p = 0.22 > 0.05)
                        No change in performance detected.
distribution_statistic/small-bin-10
                        time:   [326.84 ns 327.83 ns 328.85 ns]
                        change: [-3.1397% -2.7580% -2.3693%] (p = 0.00 < 0.05)
                        Change within noise threshold.
Found 2 outliers among 100 measurements (2.00%)
  2 (2.00%) high mild
distribution_statistic/small-bin-50
                        time:   [870.22 ns 875.35 ns 879.84 ns]
                        change: [-27.110% -26.583% -26.056%] (p = 0.00 < 0.05)
                        Performance has improved.
distribution_statistic/small-bin-100
                        time:   [1.5535 us 1.5610 us 1.5689 us]
                        change: [-32.740% -32.337% -31.946%] (p = 0.00 < 0.05)
                        Performance has improved.
Found 1 outliers among 100 measurements (1.00%)
  1 (1.00%) high mild
distribution_statistic/small-bin-200
                        time:   [2.7713 us 2.7858 us 2.7994 us]
                        change: [-45.275% -44.954% -44.640%] (p = 0.00 < 0.05)
                        Performance has improved.
distribution_statistic/small-bin-500
                        time:   [6.6219 us 6.6429 us 6.6645 us]
                        change: [-56.192% -55.736% -55.308%] (p = 0.00 < 0.05)
                        Performance has improved.
Found 6 outliers among 100 measurements (6.00%)
  5 (5.00%) high mild
  1 (1.00%) high severe
distribution_statistic/small-bin-1000
                        time:   [14.166 us 14.191 us 14.211 us]
                        change: [-62.703% -62.611% -62.505%] (p = 0.00 < 0.05)
                        Performance has improved.
distribution_statistic/large-bin-50
                        time:   [326.32 ns 329.02 ns 333.14 ns]
                        change: [-60.767% -60.546% -60.304%] (p = 0.00 < 0.05)
                        Performance has improved.
Found 10 outliers among 100 measurements (10.00%)
  1 (1.00%) low mild
  7 (7.00%) high mild
  2 (2.00%) high severe
distribution_statistic/large-bin-100
                        time:   [540.97 ns 542.61 ns 544.03 ns]
                        change: [-70.124% -69.909% -69.714%] (p = 0.00 < 0.05)
                        Performance has improved.
Found 2 outliers among 100 measurements (2.00%)
  2 (2.00%) low mild
distribution_statistic/large-bin-200
                        time:   [826.40 ns 829.33 ns 832.19 ns]
                        change: [-78.922% -78.817% -78.720%] (p = 0.00 < 0.05)
                        Performance has improved.
Found 2 outliers among 100 measurements (2.00%)
  1 (1.00%) low mild
  1 (1.00%) high mild
distribution_statistic/large-bin-500
                        time:   [1.5995 us 1.6062 us 1.6134 us]
                        change: [-84.559% -84.473% -84.385%] (p = 0.00 < 0.05)
                        Performance has improved.
Found 3 outliers among 100 measurements (3.00%)
  3 (3.00%) high mild
distribution_statistic/large-bin-1000
                        time:   [2.7991 us 2.8088 us 2.8187 us]
                        change: [-89.346% -89.304% -89.261%] (p = 0.00 < 0.05)
                        Performance has improved.
Found 1 outliers among 100 measurements (1.00%)
  1 (1.00%) high mild
```

Sometime there is a regression in `small-bin-5`, in the scale of nanoseconds, I think this is acceptable.